### PR TITLE
Apply back-off strategy for order-book API calls

### LINKED ISSUE #2280: Review button blinks with Unhandled error
Reported and related to 
https://cowservices.slack.com/archives/C0375NV72SC/p1681124839607809

From time to time I receive 503 errors in Staging (BARN API) for these requests
 - Request URL: https://barn.api.cow.fi/goerli/api/v1/quot
 - Request URL: https://barn.api.cow.fi/goerli/api/v1/token/0x91056D4A53E1faa1A84306D4deAEc71085394bC8/native_price
and it is resulting in showing 'Unhandled error' message on the UI. In 5-10 seconds requests are updated with 200 code.
Reproducible for all networks.
.

**Expected**: show a loader/other error, but not blink with 'Unhandled error'

### PR DESCRIPTION
**Pull request by shoom3301**
The strategy was accidentally lost after a major SDK update.

# Summary

Fixes #2280

  # To Test

For api.cow.fi requests should work as before: https://github.com/cowprotocol/cowswap/pull/2069

### CODE REVIEW

**Review by alfetopito — APPROVED** [review]
Not for this PR, just wondering: should we move the rate-limiting to the SDK?

**Comment by shoom3301** [comment]
@alfetopito it's a good question, let's discuss it today

**Review by anxolin — APPROVED** [review]
Soft approve with comments

**Review by elena-zh — APPROVED** [review]
Review button is no longer blinking with the error, so I'm approving!

File: src/cow-react/modules/wallet/web3-react/connection/trust.tsx [hunk 1]
```diff
@@ -22,11 +22,11 @@ const BASE_PROPS = {
...
-  new InjectedWallet({
-    actions,
-    walletUrl: WALLET_LINK,
-    searchKeywords: ['isTrust', 'isTrustWallet'],
-  })
+    new InjectedWallet({
+      actions,
+      walletUrl: WALLET_LINK,
+      searchKeywords: ['isTrust', 'isTrustWallet'],
+    })
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/wallet/web3-react/connection/trust.tsx:29]: https://github.com/cowprotocol/cowswap/pull/2198
Was introduced here

File: src/cow-react/common/utils/fetch.ts [hunk 2]
```diff
@@ -40,6 +40,26 @@ export function fetchWithRateLimit(
...
 }
 
+export function requestWithRateLimit(
+  params?: FetchWithRateLimit
+): <T>(request: () => Promise<T>) => ReturnType<typeof request> {
+  const { backoff, rateLimit } = params || {}
+
+  const limiter = rateLimit ? new RateLimiter(rateLimit) : undefined
+
+  return (request) =>
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/common/utils/fetch.ts:50]: I have tried :( And it also worries me.
However, I haven't found a better solution.
The main problem in the callback type. I will be happy to get any help with it 🙏

File: src/cow-react/api/gnosisProtocol/api.ts [hunk 3]
```diff
@@ -132,19 +137,29 @@ export async function getQuote(params: FeeQuoteParams): Promise<OrderQuoteRespon
...
+  return orderBookFetchRateLimitted(() => orderBookApi.getOrders({ owner, limit, offset }, { chainId }))
+}
+
+export async function getTrades(chainId: ChainId, owner: string): Promise<Trade[]> {
+  return orderBookFetchRateLimitted(() => orderBookApi.getTrades({ owner }, { chainId }))
+}
+
+export async function getNativePrice(chainId: ChainId, currencyAddress: string): Promise<NativePriceResponse> {
+  return orderBookFetchRateLimitted(() => orderBookApi.getNativePrice(currencyAddress, { chainId }))
 }
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/api/gnosisProtocol/api.ts:163]: Agree! Added TODO

File: src/cow-react/api/gnosisProtocol/api.ts [hunk 4]
```diff
@@ -132,19 +137,29 @@ export async function getQuote(params: FeeQuoteParams): Promise<OrderQuoteRespon
   const { chainId } = params
   const quoteParams = _mapNewToLegacyParams(params)
 
-  return orderBookApi.getQuote(quoteParams, { chainId }).catch((error) => {
+  return orderBookFetchRateLimitted(() => {
+    return orderBookApi.getQuote(quoteParams, { chainId })
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/api/gnosisProtocol/api.ts:141]: I used to use it for debugging, it's much easier to add breakpoint by clicking in a line in Chrome debugger

File: src/cow-react/common/utils/fetch.ts [hunk 5]
```diff
@@ -40,6 +40,26 @@ export function fetchWithRateLimit(
     )
 }
 
+export function requestWithRateLimit(
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/common/utils/fetch.ts]: It's not relevant enymore

**Pull request merged**